src/transaction.ts

```diff
@@ -0,0 +1,234 @@
+    inputs.length,
+    outputs.length,
+    satsPerByte,
+    network
+  );
+
+  const changeIndexLBTC: number = changeOutputs.findIndex(
+    out => out.asset === network.assetHash
+  );
+  const diff = changeOutputs[changeIndexLBTC].value - fee.value;
+
+  if (diff > 0) {
+    changeOutputs[changeIndexLBTC].value = diff;
+    const outs = outputs.concat(changeOutputs).concat(fee);
+    return addToTx(psetBase64, inputs, outs);
+  }
+  // remove the change outputs
+  changeOutputs.splice(changeIndexLBTC, 1);
+
+  if (diff === 0) {
+    const outs = outputs.concat(changeOutputs).concat(fee);
+    return addToTx(psetBase64, inputs, outs);
+  }
+
+  const availableUnspents: UtxoInterface[] = [];
+  for (const utxo of unspents) {
+    if (!selectedUtxos.includes(utxo)) availableUnspents.push(utxo);
+  }
+
+  const coinSelectionResult = greedyCoinSelection(
+    availableUnspents,
+    // a little trick to only select the difference not covered by the change output
+    [{ ...fee, value: diff }],
+    changeScriptByAsset
+  );
+
+  const ins = inputs.concat(coinSelectionResult.selectedUtxos);
+  const outs = outputs
+    .concat(changeOutputs)
+    .concat(fee)
+    .concat(coinSelectionResult.changeOutputs);
+
+  return addToTx(psetBase64, ins, outs);
+}
+
+/**
+ * select utxo for outputs among unspents.
+ * @param unspents a set of unspents.
+ * @param outputs the outputs targetted by the coin selection
+ */
+export function greedyCoinSelection(
+  unspents: UtxoInterface[],
+  outputs: OutputInterface[],
+  changeScriptGetter: ChangeScriptFromAssetGetter
+): CoinSelectionResult {
+  const result: CoinSelectionResult = {
+    selectedUtxos: [],
+    changeOutputs: [],
+  };
+
+  const utxosGroupedByAsset = groupBy(unspents, 'asset') as Record<
+    string,
+    UtxoInterface[]
+  >;
+  const outputsGroupedByAsset = groupBy(outputs, 'asset') as Record<
+    string,
+    OutputInterface[]
+  >;
+
+  for (const [asset, outputs] of Object.entries(outputsGroupedByAsset)) {
+    const unspents = utxosGroupedByAsset[asset];
+    if (!unspents) {
+      throw new Error('need unspents for the asset: ' + asset);
+    }
+
+    const targetAmount: number = outputs.reduce(
+      (acc: number, output: OutputInterface) => acc + output.value,
+      0
+    );
+
+    const { selected, changeAmount } = selectUtxos(unspents, targetAmount);
+
+    result.selectedUtxos.push(...selected);
+
+    if (changeAmount > 0) {
+      const changeScript = changeScriptGetter(asset);
+      if (!changeScript) {
+        throw new Error('need change script for asset: ' + asset);
+      }
+
+      result.changeOutputs.push({
+        asset: asset,
+        value: changeAmount,
+        script: changeScript,
+      });
+    }
+  }
+
+  return result;
+}
+
+function selectUtxos(
+  utxos: UtxoInterface[],
+  targetAmount: number
+): {
+  selected: UtxoInterface[];
+  changeAmount: number;
+} {
+  const selected: UtxoInterface[] = [];
+  let total = 0;
+  for (const utxo of utxos) {
```

Comment:
I would first sort them from the smallest till the bigger coin,

Reply:
But I would "fragment" my utxo set, making the future spends maybe bigger and also lowering my privacy, because the heuristic of the single owner of many inputs.

Reply:
Depending on the need of the user you both could be right or wrong.
Since this is devkit, we should plan to have a "pluggable" coin selection... we could create a coinselect standalone submodule that exposes an interface and some of its implementation so that it will be the user the one deciding how his utxos are going to be selected.

For now, it's ok for me to sort it and then coinselect.. privacy is more important than tx size after all.